### ISSUE TITLE
[Yul/EVM] Discussion: memory escalation of local variables.

### ISSUE THREAD

**Issue by ekpyron** [ethereum/solidity#5766]
Problem: only the last 16 stack slots are accessible. Especially since the optimizer may generate new local variables we should be able to deal with situations, in which the accessible stack height  is no longer sufficient for accessing all necessary local variables.
Solution: Escalate local variables to memory, if stack is insufficient.

Concept:

Fixed memory location ``0x80`` contains a pointer to the current "memory variable stack". It is updated on function entry and restored on function exit.
Fixed memory location ``0xA0`` points to the first entry in a linked list of free chunks of memory to be used as memory variable stack or zero, if no free chunk is available.

Convert the following:
```
function h() -> x
{
	let a_1 := 1
	let a_2 := 2
	let a_3 := 3
	let a_4 := 4
	let a_5 := 5
	let a_6 := 6
	let a_7 := 7
	let a_8 := 8
	let a_9 := 9
	let a_10 := 10
	let a_11 := 11
	let a_12 := 12
	let a_13 := 13
	let a_14 := 14
	let a_15 := 15
	let a_16 := 16
	let a_17 := 17
	let a_18 := 18
	let a_19 := 19


	let b := f(a_1, a_2, a_3, a_4, a_5)
	
	let c := g(a_6, a_7, a_8, a_9, a_10, a_11, a_12, a_13, a_14, a_15, a_16, a_17, a_18, a_19)

	x := ltu256(b, c);
}
```
to this:
```
function h2() -> x
{
	{
		// fetch "re-usable memory chunk pointer"
		let memstack := mload(0xA0)
		switch memstack {
			case 0 { // no re-usable memory chunk; allocate new chunk
				memstack := mload(0x40)
				mstore(0x40, memstack + MAX_MEMSTACK_SIZE) // MAX_MEMSTACK_SIZE is the size of the memory stack needed by the globally "worst" function
			}
			default { // we found a re-usable memory chunk
				mstore(0xA0, mload(memstack)) // mload(memstack) contains a pointer to the next free memory chunk (or 0)
			}
		}
		// now memstack points to the memory area to be used for variable escalation

		// now we store the current value of the "memory stack pointer" (0x80) to our memory chunk
		mstore(memstack, mload(0x80))
		mstore(0x80, add(memstack, 0x20)) // store our memory stack to 0x80 (add 0x20 to avoid having to offset by one later)
	}

	// now memory variable n can be found at memory address (add(mload(0x80), N)) where N is mul(0x20, n)

	// ASSIGN:	a_1 <=> mload(0x80)
	//		a_2 <=> add(mload(0x80), 0x20)
	//		a_3 <=> add(mload(0x80), 0x40)
	//		a_4 <=> add(mload(0x80), 0x60)
	//		a_5 <=> add(mload(0x80), 0x80)
	
	mstore(mload(0x80), 1) // let a_1 := 1
	mstore(add(mload(0x80), 0x20), 2) // let a_2 := 2
	mstore(add(mload(0x80), 0x40), 3) // let a_3 := 3
	mstore(add(mload(0x80), 0x60), 4) // let a_4 := 4
	mstore(add(mload(0x80), 0x80), 5) // let a_5 := 5
	let a_6 := 6
	let a_7 := 7
	let a_8 := 8
	let a_9 := 9
	let a_10 := 10
	let a_11 := 11
	let a_12 := 12
	let a_13 := 13
	let a_14 := 14
	let a_15 := 15
	let a_16 := 16
	let a_17 := 17
	let a_18 := 18
	let a_19 := 19

	// de-escalate
	{
		let a_1 := mload(mload(0x80))
		let a_2 := mload(add(mload(0x80), 0x20))
		let a_3 := mload(add(mload(0x80), 0x40))
		let a_4 := mload(add(mload(0x80), 0x60))
		let a_5 := mload(add(mload(0x80), 0x80))
		let b := f(a_1, a_2, a_3, a_4, a_5)
	}
	
	let c := g(a_6, a_7, a_8, a_9, a_10, a_11, a_12, a_13, a_14, a_15, a_16, a_17, a_18, a_19)

	x := ltu256(b, c);
	
	{
		// restore the "memory stack pointer" at 0x80
		let memchunk := sub(mload(0x80), 0x20)
		mstore(0x80, mload(memchunk))
		
		// "free" our memory chunk
		mstore(memchunk, mload(0xA0))
		mstore(0xA0, memchunk)
	}
}
```

**Comment by ekpyron** [comment]
We can extract the management code into helpers - they in fact don't even need any parameters:
```

function enterMemoryStackFrame()
{
	// fetch "re-usable memory chunk pointer"
	let memstack := mload(0xA0)
	switch memstack {
		case 0 { // no re-usable memory chunk; allocate new chunk
			memstack := mload(0x40)
			mstore(0x40, memstack + MAX_MEMSTACK_SIZE) // MAX_MEMSTACK_SIZE is the size of the memory stack needed by the "worst" function in the contract
		}
		default { // we found a re-usable memory chunk
			mstore(0xA0, mload(memstack)) // mload(memstack) contains a pointer to the next free memory chunk (or 0)
               }
	}
	// now memstack points to the memory area to be used for variable escalation

	// now we store the current value of the "memory stack pointer" (0x80) to our memory chunk
	mstore(memstack, mload(0x80))
	mstore(0x80, add(memstack, 0x20)) // store our memory stack to 0x80 (add 0x20 to avoid having to offset by one later)
}

function exitMemoryStackFrame()
{
	// restore the "memory stack pointer" at 0x80
	let memchunk := sub(mload(0x80), 0x20)
	mstore(0x80, mload(memchunk))
	
	// "free" our memory chunk
	mstore(memchunk, mload(0xA0))
	mstore(0xA0, memchunk)
}
```

Further always-to-be-inlined setters and getters:
```
function setMemVar(id, value)
{
        // the multiplication by 0x20 can be moved outside of the function
	mstore(add(mload(0x80), mul(0x20, id)), value)
}

function getMemVar(id) -> value
{
	value := mload(add(mload(0x80), mul(0x20, id)))
}
```

Yielding:

```
	function h() -> x
	{
		enterMemoryStackFrame()

		setMemVar(0, 1) // let a_1 := 1
		setMemVar(1, 2) // let a_2 := 2
		setMemVar(2, 3) // let a_3 := 3
		setMemVar(3, 4) // let a_4 := 4
		setMemVar(4, 5) // let a_5 := 5
		let a_6 := 6
		let a_7 := 7
		let a_8 := 8
		let a_9 := 9
		let a_10 := 10
		let a_11 := 11
		let a_12 := 12
		let a_13 := 13
		let a_14 := 14
		let a_15 := 15
		let a_16 := 16
		let a_17 := 17
		let a_18 := 18
		let a_19 := 19

		// de-escalate
		let b := f(getMemVar(0), getMemVar(1), getMemVar(2), getMemVar(3), getMemVar(4))
		
		let c := g(a_6, a_7, a_8, a_9, a_10, a_11, a_12, a_13, a_14, a_15, a_16, a_17, a_18, a_19)

		x := ltu256(b, c);

		exitMemoryStackFrame()
	}
```

**Comment by ekpyron** [comment]
We discussed moving the stack optimization logic out of the evm code generation. I think it makes more sense to do proper control flow and life-time analysis for local variables for Yul itself (that may be helpful in general) and then use that information for stack allocation instead (the results should be better than the current optimization during code generation).

**Comment by axic** [comment]
The main problem I find with this approach is that memory locations have a special meaning in Yul now. Perhaps we need to introduce pragmas to Yul which can define certain properties so that the compiler generating Yul can have better control over memory.

As an example:
```
pragma free_memory_pointer 0x40;
```

**Comment by ekpyron** [comment]
We could replace accessing 0x40 with ``allocate`` as already suggested in #5107 and do something similar for 0x80 and 0xA0...

**Comment by axic** [comment]
That's true, but still in general since memory pointers do not have a special type/meaning (#3525 proposes that) the compiler generating Yul can do whatever it pleases. Important to keep in mind other users past Solidity.

**Comment by ekpyron** [comment]
Yes, I see your point.

**Comment by chriseth** [comment]
Yes, this is a problem, we have to add that option to the dialect. It can only be used in Solidity if all inline assembly blocks provide a special option in the string argument to the `assembly` keyword.

**Comment by axic** [comment]
We've tried to keep memory hidden from Yul, perhaps the real solution is breaking that assumption and making memory a first class citizen of Yul.

**Comment by chriseth** [comment]
One might argue that the ABI encoder / decoder code can always use it, because it already allocates memory that way - even though not all ways to invoke the encoder and decoder do that.

**Comment by ekpyron** [comment]
All inline assembly blocks containing ``mstore`` would be sufficient.

**Comment by chriseth** [comment]
@axic yep, make (allocatable) memory a first class citizen of Yul -> #5107

**Comment by chriseth** [comment]
Is it possible to do this inside the memory object framework, so that it would be possible to optimize e.g. external functions that are never called internally to use static locations for the variables?

**Comment by ekpyron** [comment]
Just so I don't loose the link https://hackmd.io/@YuZ07BTxTK-q-lurqUdnkQ/By68AjDhL - here are some initial notes about escalating statically in the case of acyclic call graphs - I have an algorithm, but its description still can be greatly simplified.

**Comment by ekpyron** [comment]
I'll be working on a first implementation of the statically assigned memory slots for acyclic call graphs.

**Comment by ekpyron** [comment]
Starting to implement the static acyclic version described in hackmd, the following is the first question:

- Currently the IR codegen directly initializes memory as ``mstore(0x40, <constant>)``, where constant is ``generalPurposeMemoryStart`` plus ``reservedMemory`` (reserved for immutables during contract creation).
- The value ``generalPurposeMemoryStart`` is no longer known in advance.

So we probably need some kind of builtin for this. Maybe ``generalPurposeMemoryStart()`` or ``memoryInit()``. For the latter we'd need to do ``memoryInit() mstore(0x40, add(mload(0x40), <reservedMemory>)`` whereas for the former it'd just be ``mstore(0x40, add(generalPurposeMemoryStart(), <reservedMemory>))``.

I think as proof-of-concept I'll add a ``generalPurposeMemoryStart()`` builtin - but maybe there's a better way to do this.

**Comment by chriseth** [comment]
My intuition would be to make this step start with a heuristic: Search for `mstore(0x40, SOMETHING)` among the first statements of the main routine. If there is some statement that accesses memory before the `mstore`, fail. If the statement cannot be found, then fail. Otherwise, replace it by `mstore(0x40, add(<whatever we need>, SOMETHING))`.

**Comment by ekpyron** [comment]
Really :-)? Ok, I can look into that...

**Comment by cameel** [comment]
Since the feature was released in 0.8.13, I think we can close this now?
